contracts/adapters/SwapAdapter.sol

```diff
@@ -0,0 +1,468 @@
+        bytes depositDataAfterAmount;
+        bytes path;
+        bytes depositData;
+    }
+
+    error CallerNotAdmin();
+    error AlreadySet();
+    error TokenInvalid();
+    error PathInvalid();
+    error MsgValueLowerThanFee(uint256 value);
+    error InsufficientAmount(uint256 amount);
+    error FailedFundsTransfer();
+    error AmountLowerThanFee(uint256 amount);
+
+    event TokenResourceIDSet(address token, bytes32 resourceID);
+    event TokensSwapped(address tokenIn, address tokenOut, uint256 amountIn, uint256 amountOut);
+
+    /**
+        @dev The contract uses Uniswap UniversalRouter and Permit2. These addresses
+        should be set during initialization. Addresses of deployed Uniswap contracts can be
+        found in Uniswap docs.
+    */
+
+    constructor(
+        IBridge bridge,
+        address weth,
+        IUniversalRouter swapRouter,
+        IPermit2 permit2,
+        INativeTokenAdapter nativeTokenAdapter
+    ) {
+        _bridge = bridge;
+        _weth = weth;
+        _swapRouter = swapRouter;
+        _permit2 = permit2;
+        _nativeTokenAdapter = nativeTokenAdapter;
+        IERC20(_weth).approve(address(_permit2), type(uint256).max);
+        _setupRole(DEFAULT_ADMIN_ROLE, msg.sender);
+    }
+
+    modifier onlyAdmin() {
+        if (!hasRole(DEFAULT_ADMIN_ROLE, msg.sender)) revert CallerNotAdmin();
+        _;
+    }
+
+    // Admin functions
+    function setTokenResourceID(address token, bytes32 resourceID) external onlyAdmin {
+        if (tokenToResourceID[token] == resourceID) revert AlreadySet();
+        tokenToResourceID[token] = resourceID;
+        IERC20(token).approve(address(_permit2), type(uint256).max);
+        emit TokenResourceIDSet(token, resourceID);
+    }
+
+    /**
+        @notice Function for depositing tokens, performing swap to ETH and bridging the ETH.
+        @param destinationDomainID  ID of chain deposit will be bridged to.
+        @param recipient Recipient of the deposit.
+        @param token Input token to be swapped.
+        @param amountInMax Max amount of input tokens to be swapped. It should exceed the desired
+            amount of output tokens because the amount of swapped ETH should also cover the bridging fee.
+            It's equal to tokenSwapRate * (amountOut + bridging fee)
+        @param amountOut Amount of ETH to be bridged.
+        @param pathTokens Addresses of the tokens for Uniswap swap (in reverse order). WETH address is used for ETH.
+        @param pathFees Fees for Uniswap pools (in reverse order).
+    */
+    function depositTokensToEth(
+        uint8 destinationDomainID,
+        address recipient,
+        address token,
+        uint256 amountInMax,
+        uint256 amountOut,
+        address[] calldata pathTokens,
+        uint24[] calldata pathFees
+    ) external {
+        LocalVars memory vars;
+        vars.resourceID = tokenToResourceID[token];
+        if (vars.resourceID == bytes32(0)) revert TokenInvalid();
```

Comment:
This is wrong?! resourceId in this method should always be resourceId of native Eth. 
Because for this to work we would have to set `tokenToResourceID[usdc] = nativeEthResourceId` which would then screw up `depositEthToTokens` where the same call needs usdc resource id

Reply:
Correct, we always relay ETH in this case, so recourceID should be constant.